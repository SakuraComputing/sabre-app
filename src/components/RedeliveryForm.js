import React from 'react';
import moment from 'moment';

export default class RedeliveryForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            id: props.redelivery ? props.redelivery.id : '',
            cardId: props.redelivery ? props.redelivery.cardId : '',
            parcelId: props.redelivery ? props.redelivery.parcelId : '',
            latestScan: props.redelivery ? moment(props.redelivery.latestScan).format('YYYY-MM-DD') : undefined,
            requestDate: props.redelivery ? moment(props.redelivery.requestDate).format('YYYY-MM-DD') : undefined,
            error: ''
        }
    }

    onCardIdChange = (e) => {
      const cardId = e.target.value;
      if (!cardId || cardId.match(/\d/)) {
          this.setState(() => ({ cardId }));
      };
    };
    onParcelIdChange = (e) => {
      const parcelId = e.target.value;
      this.setState(() => ({ parcelId }));
    };
    onLatestDateChange = (e) => {
      const latestScan = e.target.value;
      this.setState(() => ({ latestScan }));
    };

    onRequestDateChange = (e) => {
        const requestDate = e.target.value;
        this.setState(() => ({ requestDate }));
    };

    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.cardId && !this.parcelId) {
            this.setState(() => ({ error: 'A parcel or card id must be input'}));
        }  else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                id: this.state.id,
               cardId: this.state.cardId,
               parcelId: this.state.parcelId,
               latestScan: this.state.latestScan,
               requestDate: this.state.requestDate
            });
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <div>{this.state.error}</div>}
                <form onSubmit={this.onSubmit} >
                    <input
                        type="text"
                        className="card-id"
                        value={this.state.cardId}
                        placeholder="Card Id"
                        onChange={this.onCardIdChange}
                    />
                    <input
                        type="text"
                        className="parcel-id"
                        value={this.state.parcelId}
                        placeholder="Parcel Id"
                        onChange={this.onParcelIdChange}
                    />
                    <input
                        type="date"
                        className="latest-scan"
                        value={this.state.latestScan}
                        placeholder="Latest Scan"
                        onChange={this.onLatestDateChange}
                    />
                    <input
                        type="date"
                        className="request-date"
                        value={this.state.requestDate}
                        placeholder="Request Date"
                        onChange={this.onRequestDateChange}
                    />
                    <button>Save Redelivery</button>
                </form>
            </div>
        )
    }

}
