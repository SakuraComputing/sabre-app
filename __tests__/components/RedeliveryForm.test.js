import React from 'react';
import RedeliveryForm from '../../src/components/RedeliveryForm';
import { shallow } from 'enzyme';
import redeliveries from '../fixtures/redeliveries';
import moment from 'moment';

describe('Redelivery Form',() => {
    describe('Form inputs',() => {

        let wrapper;

        beforeEach(() => {
            wrapper = shallow(<RedeliveryForm/>);
        });
        it('should render the form correctly',() => {
            expect(wrapper).toMatchSnapshot();
        });
        it('should set card id on input form', () => {
            const value = '888888';
            wrapper.find('.card-id').simulate('change',
                { target: { value }
                });
            expect(wrapper.state('cardId')).toBe(value);
        });
        it('should not set a card id for and incorrect input', () => {
            const cardId = 'WRONG';
            wrapper.find('.card-id').simulate('change',
                { target: { cardId }
                });
            expect(wrapper.state('cardId')).toBeUndefined();
        });
        it('should set parcel id on input form', () => {
            const value = 'JD888888';
            wrapper.find('.parcel-id').simulate('change',
                { target: { value }
                });
            expect(wrapper.state('parcelId')).toBe(value);
        });
        it('should set latest scan', ()=> {
            const now = moment().format('YYYY-MM-DD');
            wrapper.find('.latest-scan').simulate('change', { target: { value: now }});
            expect(wrapper.state('latestScan')).toEqual( now );
        });
        it('should set the request date',()=> {
            const now = moment().format('YYYY-MM-DD');
            wrapper.find('.request-date').simulate("change", { target: { value: now }});
            expect(wrapper.state('requestDate')).toEqual( now );
        });
    });

    it('should render the form with data',() => {
        const wrapper = shallow(<RedeliveryForm redelivery={redeliveries[0]}/>);
        expect(wrapper).toMatchSnapshot();
    });

    describe('when the form is submitted',() => {
        it('should render an error when for an invalid form submission',() => {
            const wrapper = shallow(<RedeliveryForm />);
            wrapper.find('form').simulate('submit', {
                preventDefault: () => {}
            });
            expect(wrapper.state('error').length).toBeGreaterThan(0);
            expect(wrapper).toMatchSnapshot();
        });
        it('should submit props for valid form submission',() => {
            const onSubmitSpy = jest.fn();
            const wrapper = shallow(<RedeliveryForm redelivery={redeliveries[0]} onSubmit={onSubmitSpy}/>)
            wrapper.find('form').simulate('submit', {
                preventDefault: () => {}
            });
            expect(wrapper.state('error')).toBe('');
            expect(onSubmitSpy).toHaveBeenCalledWith({
                cardId: redeliveries[0].cardId,
                id: redeliveries[0].id,
                parcelId: redeliveries[0].parcelId,
                latestScan: moment(redeliveries[0].latestScan).format('YYYY-MM-DD'),
                requestDate: moment(redeliveries[0].requestDate).format('YYYY-MM-DD')
            });
        });
    });

});


