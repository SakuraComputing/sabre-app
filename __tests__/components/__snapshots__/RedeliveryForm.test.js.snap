// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redelivery Form Form inputs should render the form correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      className="card-id"
      onChange={[Function]}
      placeholder="Card Id"
      type="text"
      value=""
    />
    <input
      className="parcel-id"
      onChange={[Function]}
      placeholder="Parcel Id"
      type="text"
      value=""
    />
    <input
      className="latest-scan"
      onChange={[Function]}
      placeholder="Latest Scan"
      type="date"
    />
    <input
      className="request-date"
      onChange={[Function]}
      placeholder="Request Date"
      type="date"
    />
    <button>
      Save Redelivery
    </button>
  </form>
</div>
`;

exports[`Redelivery Form should render the form with data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      className="card-id"
      onChange={[Function]}
      placeholder="Card Id"
      type="text"
      value="876363"
    />
    <input
      className="parcel-id"
      onChange={[Function]}
      placeholder="Parcel Id"
      type="text"
      value="JD0002289756473452"
    />
    <input
      className="latest-scan"
      onChange={[Function]}
      placeholder="Latest Scan"
      type="date"
      value="2010-01-01"
    />
    <input
      className="request-date"
      onChange={[Function]}
      placeholder="Request Date"
      type="date"
      value="2010-01-01"
    />
    <button>
      Save Redelivery
    </button>
  </form>
</div>
`;

exports[`Redelivery Form when the form is submitted should render an error when for an invalid form submission 1`] = `
<div>
  <div>
    A parcel or card id must be input
  </div>
  <form
    onSubmit={[Function]}
  >
    <input
      className="card-id"
      onChange={[Function]}
      placeholder="Card Id"
      type="text"
      value=""
    />
    <input
      className="parcel-id"
      onChange={[Function]}
      placeholder="Parcel Id"
      type="text"
      value=""
    />
    <input
      className="latest-scan"
      onChange={[Function]}
      placeholder="Latest Scan"
      type="date"
    />
    <input
      className="request-date"
      onChange={[Function]}
      placeholder="Request Date"
      type="date"
    />
    <button>
      Save Redelivery
    </button>
  </form>
</div>
`;
